Integrate task resource (#237)
iterative/terraform-provider-iterative: docs/index.md

# Iterative Provider

Use the Iterative Provider to launch resource-intensive tasks in popular cloud

> providers with just a simple Terraform file.

Reviewer: p.s. I noticed docs/guides/getting-started.md is not wrapped at 80 chars. Seems inconsistent 💅

Reviewer: BTW, minor but this can be read as condescending.
```suggestion
Use the Iterative Provider to launch resource-intensive tasks in popular cloud
providers with a single Terraform file.
```

## Example Usage

```hcl

> terraform {

Reviewer: This section shows a sample config but not it's usage. Should we at least also mention `terraform init/apply`?
And/or link to the full Get Started guide.

Reviewer: p.s. this may apply to docs/resources/task.md additionally (or instead).

Author: Popular providers' documentation uses the `index.md` **Example Usage** section to target those users that know how to use Terraform and just need a bit of boilerplate to get started.

New users should follow the **Getting Started** guide instead. Still, we have code snippets in three different places; not sure if it's a good idea.

Author: https://github.com/iterative/terraform-provider-iterative/pull/237/files/739955bf1943bb25a1d3711ce232f4d5ff67676b#r749787219

  required_providers {
    iterative = {
      source  = "iterative/iterative"
    }
  }
}

provider "iterative" {}

resource "iterative_task" "task" {
  name  = "example"
  cloud = "aws"

  script = <<-END
    #!/bin/bash
    echo "hello!"
  END
}
```

> 

Author: -> **Note:** See the [Getting Started](https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/getting-started) for more information.

## Authentication

Environment variables are the only supported authentication method. They should

> be present when running any of the `terraform` commands.

Reviewer: Prob link to https://registry.terraform.io/providers/iterative/iterative/latest/docs#authentication from here too?

Author: ```suggestion
[Environment variables](https://registry.terraform.io/providers/iterative/iterative/latest/docs#authentication) are the only supported authentication method. They should
be present when running any of the `terraform` commands.
```

### Example

```bash
$ export GOOGLE_APPLICATION_CREDENTIALS_DATA="$(cat service_account.json)"
$ terraform apply
```

### Amazon Web Services

- `AWS_ACCESS_KEY_ID` - Access key identifier.
- `AWS_SECRET_ACCESS_KEY` - Secret access key.
- `AWS_SESSION_TOKEN` - (Optional) Session token.

### Google Cloud Platform

- `GOOGLE_APPLICATION_CREDENTIALS` - Path to a service account JSON key file.

-> **Note** you can also use `GOOGLE_APPLICATION_CREDENTIALS_DATA` with the
**contents** of the service account JSON key file.

### Microsoft Azure

- `AZURE_CLIENT_ID` - Client identifier.
- `AZURE_CLIENT_SECRET` - Client secret.
- `AZURE_SUBSCRIPTION_ID` - Subscription identifier.
- `AZURE_TENANT_ID` - Tenant identifier.

### Kubernetes

- `KUBECONFIG` - Path to a [`kubeconfig` file](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#the-kubeconfig-environment-variable).

> -> **Note** you can also use `KUBECONFIG_DATA` with the **contents** of the `kubeconfig` file.

Reviewer: ```suggestion
> **Note**: You can use `KUBECONFIG_DATA` instead, with the **contents** of the `kubeconfig` file.
```

## Argument Reference

> This module doesn't have any top-level arguments. See the [task resource](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task) for more information.

Reviewer: Unclear note. What module? Why would we expect arguments?
Maybe it's not needed at all?

Could link to the task resource ref. from the Usage block instead, or even from the intro.

Author: > _Why would we expect arguments?_

In a future, we may want to accept secrets and some configuration values through module arguments instead of environment variables. Still, this won't happen before merging this pull request, so I'm deleting the section and adding this comment to https://github.com/iterative/terraform-provider-iterative/pull/237#issue-776178285

Author: ```suggestion
```
